File: examples/python/qfssample.py
#
# $Id$
#
# Author: Thilee Subramaniam
#
# Copyright 2012,2016 Quantcast Corporation. All rights reserved.
#
# This file is part of Kosmos File System (KFS).
#
# Licensed under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied. See the License for the specific language governing
# permissions and limitations under the License.

#
# NOTE: The python support for QFS is EXPERIMENTAL at this stage. The
#       python extension module has not been tested on large scale
#       deploymentsi yet. Please excercise caution while using the
#       python module.

"""
This simple test tries to create some files and directories, and write some
data at specific offsets in the created files. Then it tries to ensure that
the created paths are valid, and that the file contents are as expected.

To run this script,
  - Prepare qfs.so as described in the file 'doc/ClientDeveloperDoc'
  - Ensure that the QFS metaserver and chunkserver are running.
  - Ensure that the metaserver host/port matches the contents of argv[1].
  - Ensure that the PYTHONPATH and LD_LIBRARY_PATH are set accordingly.
  eg: PYTHONPATH=${PYTHONPATH}:~/code/qfs/build/lib/lib64/python \
      LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:~/code/qfs/build/lib    \
      python ./qfssample.py qfssample.cfg
"""

import os
import sys
import time

import qfs


def ParseConfig(config):
    host = ""
    port = -1
    for line in open(config):
        if line.startswith("#") or len(line.strip()) == 0:
            continue
        s = line.strip()
        if s.split("=")[0].strip() == "metaServer.name":
            host = s.split("=")[1].strip()
        elif s.split("=")[0].strip() == "metaServer.port":
            port = int(s.split("=")[1].strip())
    if (host, port) == ("", -1):
        err_exit("Failed to parse config file")
    return (host, port)


def err_exit(msg):
    print(msg)
    sys.exit(1)


def main():
    if len(sys.argv) < 2:
        err_exit("Usage: %s config_file" % sys.argv[0])

    client = None
    server = ParseConfig(sys.argv[1])

    # f = open("ztest", "wt", encoding="utf-8", errors="ignore")
    # print("encoding: " + f.encoding)
    # print("mode: " + f.mode)
    # print("file: " + str(f))
    # return 0
    try:
        client = qfs.client(server)
    except:
        err_exit(
            "Unable to start the QFS client.\n"
            + "Make sure that the meta- and chunkservers are running."
        )

    print("qfs client: " + str(client))
    client.log_level("DEBUG")
    client.log_level("ERROR")

    testBaseDir = "qfssample_base"
    testDirs = ("dir1", "dir2")
    testFile1 = "dir1/file1"
    testFile2 = "file2"
    file1Content = "Cu populo nusquam alienum vim, graece latine prodesset ex qui, quo ea lucilius intellegat."
    file2ContentA = {
        0: "are ",  # at offset 0
        40: "you ",  # at offset 40
        1030: "always ",
        1048580: "wrong?",
    }
    file2ContentB = {500: "really "}

    client.cd("/")

    try:  # just in case we didn't cleanup last time
        client.rmdirs(testBaseDir)
    except IOError as err:
        pass

    client.mkdir(testBaseDir)
    client.cd(testBaseDir)
    for td in testDirs:
        client.mkdir(td)
    time.sleep(1)
    print("Created directories.")

    client.cd("/" + testBaseDir)
    f1 = client.create(testFile1, 2)
    f2 = client.create(testFile2, 3)

    f1.write(file1Content)
    for offset, content in list(file2ContentA.items()):
        f2.seek(offset)
        f2.write(content)
    print("Created files.")

    f1.sync()
    f1.close()
    f2.sync()
    f2.close()
    time.sleep(1)
    print("Closed files (first time).")

    f1 = client.open(testFile1, "r", "utf-8", "ignore")
    f2 = client.open(testFile2, "w")
    print("Opened files.")

    for offset, content in list(file2ContentB.items()):
        f2.seek(offset)
        f2.write(content)

    f1.sync()
    f1.close()
    f2.sync()
    f2.close()
    time.sleep(1)
    print("Closed files (second time).")

    # Verify if everything is fine.
    client.cd("/")
    expected = ("dir1", "dir2", "file2")
    for node in client.readdir(testBaseDir):
        print(node)
        if node in (".", ".."):
            continue
        if node not in expected:
            err_exit("%s is not in expected list %r" % (node, expected))

    expected = "file1"
    for node in client.readdir(testBaseDir + "/dir1"):
        print(node)
        if node in (".", ".."):
            continue
        if node not in expected:
            err_exit("%s is not in expected list %r" % (node, expected))
        print("Created paths are in order.")

    print("readdirpus: " + str(client.readdirplus(testBaseDir + "/dir1")))

    filePath1 = testBaseDir + "/" + testFile1
    filePath2 = testBaseDir + "/" + testFile2

    print("Stat for %s is %r" % (filePath1, client.stat(filePath1)))
    print("Stat for %s is %r" % (filePath2, client.stat(filePath2)))

    f1 = client.open(filePath1, "r")
    out = f1.read(2)
    if out != "Cu":
        err_exit("Error: Expected 'Cu', got %r." % out)
    f1.seek(31)
    out = f1.read(6)
    if out != "graece":
        err_exit("Error: Expected 'graece', got '%s'." % out)
    pos = f1.tell()
    if pos != 37:
        err_exit("Error: Expected 'pos = 37', got 'pos = %d'." % pos)
    f1.close()
    print("File1 contents are in order")

    f2 = client.open(filePath2, "r")
    f2.seek(1032)
    out = f2.read(3)
    if out != "way":
        err_exit("Error: Expected 'way', got %r." % out)
    f2.seek(1048578)
    out = f2.read(7)
    if out[2:] != "wrong":
        err_exit("Error: Expected '..wrong', got %r." % out)
    f2.close()
    print("File2 contents are in order")

    f2 = client.open(filePath2, "r", "utf-8", "ignore")
    res = f2.data_verify()
    print("data verify %s: %d" % (filePath2, res))
    locs = f2.chunk_locations(0)
    print("chunk locations %s: %s" % (filePath2, str(locs)))
    print("file: " + str(f2))
    locs = f2.chunk_locations(0)
    print("chunk locations: " + str(locs))
    f2.close()
    f2.open("wb+")
    f2.seek(0, os.SEEK_END)
    pos = f2.tell()
    print("file end: %d" % pos)
    if 1048586 != pos:
        err_exit("Error: tell expected %d, got %d." % (1048586, out))
    f2.truncate(0)
    print("file: " + str(f2))
    f2.close()
    st = client.stat(filePath2)
    if st[6] != 0:
        err_exit("Error: truncate / stateexpected 0, got %d." % st[6])
    dst = filePath2 + "x"
    client.rename(filePath2, dst)
    client.remove(dst)
    try:
        client.remove(dst)
        dst = None
    except IOError as ex:
        print("remove " + str(ex))
    if dst is None:
        err_exit("Error: remove failed to raised exception.")
    nc = client.getNumChunks(filePath1)
    print("%s chunks: %s" % (filePath1, str(nc)))
    if nc < 1:
        err_exit("Error: chunks %d >= 1." % nc)

    sz = client.getChunkSize(filePath1)
    exp_sz = 64 << 20
    if sz != exp_sz:
        err_exit("Error: chunk size %d != %d." % (sz % exp_sz))

    if not client.isfile(filePath1):
        err_exit("Error: isfile: " + filePath1)

    if not client.isdir(testBaseDir):
        err_exit("Error: isdir: " + testBaseDir)

    stat = client.stat(filePath1)
    print("full stat %s: %s" % (filePath1, str(stat)))
    fstat = client.fullstat(filePath1)
    print("full full stat %s: %s" % (filePath1, str(fstat)))

    client.rmdirs(testBaseDir)


if __name__ == "__main__":
    main()

File: examples/java/QfsSample.java
/**
 * $Id$
 *
 * Created 2007/08/25
 * Author: Sriram Rao
 *
 * Copyright 2012 Quantcast Corp.
 * Copyright 2007 Kosmix Corp.
 *
 * This file is part of Kosmos File System (KFS).
 *
 * Licensed under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 *
 * \brief A Sample Java program to access QFSAccess APIs. To run this program,
 *        you need:
 *        - qfs-access jar in your CLASSPATH
 *        - libqfs_access.so in your LD_LIBRARY_PATH
 *        - a QFS deployment
 * Eg:
 *    javac -cp ~/code/qfs/build/java/qfs-access-<ver>.jar QfsSample.java
 *    LD_LIBRARY_PATH=~/code/qfs/build/lib \
 *        java -cp .:~/code/qfs/build/java/qfs-access-<ver>.jar QfsSample \
 *        localhost 20000
 */


import java.io.*;
import java.net.*;
import java.util.Random;
import java.nio.ByteBuffer;
import com.quantcast.qfs.access.*;

public class QfsSample
{
    public static void main(String args[]) {
        if (args.length < 1) {
            System.out.println("Usage: QfsSample <meta server> <port>");
            System.exit(1);
        }
        try {
            // From the command line, get the location of the meta-server
            int port = Integer.parseInt(args[1].trim());

            // Initialize a KfsAccess object.  The KfsAccess object is
            // the glue code that gets us into KFS world.
            //
            KfsAccess kfsAccess = new KfsAccess(args[0], port);

            String basedir = new String("jtest");

            // Make a test directory where we can do something
            if (!kfsAccess.kfs_exists(basedir)) {
                if (kfsAccess.kfs_mkdirs(basedir) != 0) {
                    System.out.println("Unable to mkdir");
                    System.exit(1);
                }
            }

            // What we just created better be a directory
            if (!kfsAccess.kfs_isDirectory(basedir)) {
                System.out.println("KFS doesn't think " + basedir + " is a dir!");
                System.exit(1);

            }

            // Create a simple file with default replication (at most 3)
            String path = new String(basedir + "/foo.1");
            KfsOutputChannel outputChannel;

            // outputChannel implements a WriteableChannel interface;
            // it is our handle to subsequent I/O on the file.
            if ((outputChannel = kfsAccess.kfs_create(path)) == null) {
                System.out.println("Unable to call create");
                System.exit(1);
            }

            // Get the directory listings
            String [] entries;
            if ((entries = kfsAccess.kfs_readdir(basedir)) == null) {
                System.out.println("Readdir failed");
                System.exit(1);
            }

            System.out.println("Readdir returned: ");
            for (int i = 0; i < entries.length; i++) {
                System.out.println(entries[i]);
            }

            // write something to the file
            int numBytes = 2048;
            char [] dataBuf = new char[numBytes];

            generateData(dataBuf, numBytes);

            String s = new String(dataBuf);
            byte[] buf = s.getBytes();

            int res = outputChannel.write(ByteBuffer.wrap(buf, 0, buf.length));
            if (res != buf.length) {
                System.out.println("Was able to write only: " + res);
            }

            // flush out the changes
            outputChannel.sync();

            // Close the file-handle
            outputChannel.close();

            // Determine the file-size
            long sz = kfsAccess.kfs_filesize(path);

            if (sz != buf.length) {
                System.out.println("System thinks the file's size is: " + sz);
            }

            // rename the file
            String npath = new String(basedir + "/foo.2");
            kfsAccess.kfs_rename(path, npath);

            if (kfsAccess.kfs_exists(path)) {
                System.out.println(path + " still exists after rename!");
                System.exit(1);
            }

            KfsOutputChannel outputChannel1 = kfsAccess.kfs_create(path);
            if (outputChannel1 != null) {
                outputChannel1.close();
            }

            if (!kfsAccess.kfs_exists(path)) {
                System.out.println(path + " doesn't exist");
                System.exit(1);
            }

            // try to rename and don't allow overwrite
            if (kfsAccess.kfs_rename(npath, path, false) == 0) {
                System.out.println("Rename with overwrite disabled succeeded!");
                System.exit(1);
            }

            // Remove the file
            kfsAccess.kfs_remove(path);

            // Verify that it is gone
            if (!kfsAccess.kfs_isFile(npath)) {
                System.out.println(npath + " is not a normal file!");
                System.exit(1);
            }

            // Re-open the file to read something.  For reads/writes,
            // Kfs provides a readable/writeable byte channel interface.
            KfsInputChannel inputChannel = kfsAccess.kfs_open(npath);
            if (inputChannel == null) {
                System.out.println("open on " + npath + "failed!");
                System.exit(1);
            }

            // read some bytes
            buf = new byte[128];
            res = inputChannel.read(ByteBuffer.wrap(buf, 0, 128));

            // Verify what we read matches what we wrote
            s = new String(buf);
            for (int i = 0; i < 128; i++) {
                if (dataBuf[i] != s.charAt(i)) {
                    System.out.println("Data mismatch at char: " + i);
                }
            }

            // seek to offset 40.  The KfsInputChannel allows seeking;
            // this is an extension to the basic readablebytechannel api.
            inputChannel.seek(40);

            // Seek and verify that we are we think we are
            sz = inputChannel.tell();
            if (sz != 40) {
                System.out.println("After seek, we are at: " + sz);
            }

            inputChannel.close();

            // remove the file
            kfsAccess.kfs_remove(npath);

            // remove the dir
            if (kfsAccess.kfs_rmdir(basedir) < 0) {
                System.out.println("unable to remove: " + basedir);
                System.exit(1);
            }
            System.out.println("All done...Test passed!");

        } catch (Exception e) {
            e.printStackTrace();
            System.out.println("Unable to setup KfsAccess");
            System.exit(1);
        }
    }

    private static Random randGen = new Random(100);

    private static void generateData(char buf[], int numBytes)
    {
        int i;

        for (i = 0; i < numBytes; i++) {
            buf[i] = (char) ('a' + (randGen.nextInt(26)));
        }
    }

}
